### PR TITLE
Python bindings for Lc0 backend.

### PR DESCRIPTION
**Pull request by mooskagh**
Building:

1. Run meson.py with -Dpython_bindings
2. After the build the file `backends.cpython-38-x86_64-linux-gnu.so` will appear in the build directory (in Windows, the extension should be .pyd).
3. Put this file in the following directory structure:
```
(root)
|
+-- lczero/
    |
    +-- __init__.py
    +-- backends.cpython-38-x86_64-linux-gnu.so
```
4. cd into (root) and start python.
5. Do `import lczero.backends` and it should all work!

### CODE REVIEW

**Comment by mooskagh** [comment]
Example session:
```

In [1]: from lczero.backends import Weights, Backend, GameState

In [2]: w = Weights()
Found pb network file: ./id58612

In [3]: w.filters()
Out[3]: 128

In [4]: Backend.available_backends()
Out[4]:
['cudnn-auto',
 'cudnn',
 'cudnn-fp16',
 'opencl',
 'blas',
 'eigen',
 'random',
 'check',
 'roundrobin',
 'multiplexing',
 'demux']

In [5]: b = Backend(weights=w)
Creating backend [cudnn-auto]...
Switching to [cudnn]...
CUDA Runtime version: 10.2.0
Cudnn version: 7.6.5
Latest version of CUDA supported by the driver: 10.2.0
GPU: GeForce GTX 970
GPU memory: 3.93951 Gb
GPU clock frequency: 1253 MHz
GPU compute capability: 5.2

In [6]: g = GameState(moves=['e2e4, e7e5'])
---------------------------------------------------------------------------
LczeroException                           Traceback (most recent call last)
<ipython-input-6-cb5a9b68a873> in <module>
----> 1 g = GameState(moves=['e2e4, e7e5'])

LczeroException: Bad move: e2e4, e7e5

In [7]: g = GameState(moves=['e2e4', 'e7e5'])

In [8]: print(g.as_string())
rnbqkbnr
pppp.ppp
........
....p...
....P...
........
PPPP.PPP
RNBQKBNR KQkq[ah] (from white's eyes) Hash: 8572518730975593504


In [9]: i1 = g.as_input(b)

In [10]: i2 = GameState(fen='2R5/5kpp/4p3/p4p2/3B4/1K5N/4rNPP/8 b - - 0 29').as_input()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-3203972071a3> in <module>
----> 1 i2 = GameState(fen='2R5/5kpp/4p3/p4p2/3B4/1K5N/4rNPP/8 b - - 0 29').as_input()

TypeError: function missing required argument 'backend' (pos 1)

In [11]: i2 = GameState(fen='2R5/5kpp/4p3/p4p2/3B4/1K5N/4rNPP/8 b - - 0 29').as_input(b)

In [12]: bin(i1.mask(0))
Out[12]: '0b10000000000001110111100000000'

In [13]: bin(i2.mask(0))
Out[13]: '0b100001000100001100000000000000'

In [14]: o1, o2 = b.evaluate(i1, i2)

In [15]: o1.q()
Out[15]: 0.10548576712608337

In [16]: o2.q()
Out[16]: -0.9555819034576416

In [17]: zip(p1.moves(), o1.p_softmax(*p1.policy_indices()))
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-17-6c83ce4b9dff> in <module>
----> 1 zip(p1.moves(), o1.p_softmax(*p1.policy_indices()))

NameError: name 'p1' is not defined

In [18]: zip(g.moves(), o1.p_softmax(*g.policy_indices()))
Out[18]: <zip at 0x7efca0e8c340>

In [19]: list(zip(g.moves(), o1.p_softmax(*g.policy_indices())))
Out[19]:
[('b1a3', 0.0020779985934495926),
 ('b1c3', 0.047005292028188705),
 ('d1e2', 0.0033786550629884005),
 ('d1f3', 0.004004619084298611),
 ('d1g4', 0.001967670163139701),
 ('d1h5', 0.004279660061001778),
 ('e1e2', 0.0014249634696170688),
 ('f1e2', 0.007004433311522007),
 ('f1d3', 0.005651619751006365),
 ('f1c4', 0.017071455717086792),
 ('f1b5', 0.0045316447503864765),
 ('f1a6', 0.001234786119312048),
 ('g1e2', 0.007879829034209251),
 ('g1f3', 0.8301827907562256),
 ('g1h3', 0.002097159158438444),
 ('a2a3', 0.00787483248859644),
 ('a2a4', 0.004778245929628611),
 ('b2b3', 0.002658447250723839),
 ('b2b4', 0.0014852079330012202),
 ('c2c3', 0.0037954614963382483),
 ('c2c4', 0.00382616207934916),
 ('d2d3', 0.006489641964435577),
 ('d2d4', 0.006822879426181316),
 ('f2f3', 0.0023090150207281113),
 ('f2f4', 0.003144552232697606),
 ('g2g3', 0.0036341759841889143),
 ('g2g4', 0.0014319035690277815),
 ('h2h3', 0.009356374852359295),
 ('h2h4', 0.0026004489045590162)]

In [20]:
```

File: meson.build [hunk 1]
```diff
@@ -543,3 +544,25 @@ if get_option('gtest')
...
+
+
+#############################################################################
+## Python bindings
+#############################################################################
+
+if get_option('python_bindings')
+  pymod = import('python')
+  python = pymod.find_installation('python3')
+  py_bindings_generator = find_program('scripts/gen_py_bindings.py')
```
Comment (borg323) [inline meson.build:556]: Add something like
```
if python.language_version() < '3.7'
  error('You need python 3.7 or newer')
endif
```

File: scripts/gen_py_bindings.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
+#!/bin/env python3
```
Comment (borg323) [inline scripts/gen_py_bindings.py:1]: `/usr/bin/env`

File: src/python/setup.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/env python3
```
Comment (borg323) [inline src/python/setup.py:1]: Same here.

Comment (mooskagh) [reply src/python/setup.py]: Actually, this file should not be there, removed.

File: meson.build [hunk 4]
```diff
@@ -543,3 +544,28 @@ if get_option('gtest')
...
+
+  gen_py_bindings = custom_target('backends', input:[], output:['backends.cc'],
+    command : [py_bindings_generator, '@OUTPUT0@'])
+
+  py_files = [ gen_py_bindings ]
+
+  cpython = dependency('python3')
+  python.extension_module('backends',
+    [py_files + files],
+    include_directories: [includes],
```
Comment (borg323) [inline meson.build:569]: Adding `cpp_args: cc.get_supported_arguments(['-Wno-c99-extensions', '-Wno-pedantic', '-Wno-missing-field-initializers']),` here silences a lot of the warnings, but probably not a good idea.

Comment (mooskagh) [reply meson.build:569]: c99-extension that I use will appear in C++20, so I'd ignore that. Other ones probably worth fixing, but I'll do it later (clang also emits some set of warnings).

**Comment by sagebei** [comment]
How to use run MCTS simulations in the backends?

**Comment by donaldlin30** [comment]
Is there a way I can retrieve centipawn scores from the binding?

**Comment by XintianHan** [comment]
How do I know the winner of the game?

**Comment by Hidancloud** [comment]
Can you guys please work a bit more with documentation and tutorials? I'm a big fan of Lc0 and there are so many great things in this repo, but there is no more or less full tutorial & explanation how and what to use here. I suppose, your solution far exceeded the amount of work that should be documented.

**Comment by psktam** [comment]
@Hidancloud not sure if you already saw this but just in case looks like there's a more up-to-date solution here: https://github.com/LeelaChessZero/lc0/pull/1856

If all you want to do is install the python bindings, looks like you just need to `cd` into the project root for this repository and run `pip install --user .` which will allow you to import the `lczero.backends` package in python.

**Pull request merged**